internal/sparsefile/sparsefile.go

```diff
@@ -8,49 +8,20 @@ import (
 	"github.com/pkg/errors"
 
 	"github.com/kopia/kopia/internal/iocopy"
-	"github.com/kopia/kopia/internal/stat"
 )
 
-// Write writes the contents of src to the given targetPath, omitting any holes.
-func Write(targetPath string, src io.Reader, size int64) error {
-	dst, err := os.OpenFile(targetPath, os.O_RDWR|os.O_CREATE|os.O_TRUNC, 0o600) //nolint:gosec,gomnd
-	if err != nil {
-		return err //nolint:wrapcheck
-	}
-
-	// ensure we always close f. Note that this does not conflict with the
-	// close below, as close is idempotent.
-	defer dst.Close() //nolint:errcheck,gosec
-
-	if err = dst.Truncate(size); err != nil {
-		return errors.Wrap(err, "error writing sparse file")
-	}
-
-	s, err := stat.GetBlockSize(targetPath)
-	if err != nil {
-		return errors.Wrap(err, "error writing sparse file")
-	}
-
+// Copy copies a file sparsely (omitting holes) from src to dst, while recycling
+// shared buffers.
+func Copy(dst io.WriteSeeker, src io.Reader) (int64, error) {
 	buf := iocopy.GetBuffer()
 	defer iocopy.ReleaseBuffer(buf)
 
-	w, err := copySparse(dst, src, buf[0:s])
-	if err != nil {
-		return errors.Wrap(err, "error writing sparse file")
-	}
-
-	if w != size {
-		return errors.Errorf("")
-	}
-
-	if err := dst.Close(); err != nil {
-		return err //nolint:wrapcheck
-	}
-
-	return nil
+	return copyBuffer(dst, src, buf)
```

Comment:
I realize now one thing missing here is we pass the entire buffer to `copyBuffer` (64K), not the intended disk block size (usually `512B`). This makes the current implementation of E2E test fail.

The next effect of this is that zero block detection/seeking will happen at that higher granularity, which is less than ideal. On the other hand, we can now enable sparse restores for Windows without much work, as we will no longer need to implement `GetDiskBlockSize()`–something that proved to be tricky for Windows.

@Shrekster we had a discussion earlier about this and I think we agreed this is an okay tradeoff. @jkowalski What do you think?

Reply:
We can also just hardcode `512B` in...

Reply:
Resolved. Got around this by simply using a closure. Windows still not enabled, I'd like to work on that in a future PR =)